File: GreedyMethod/python/fractionalKnapsack.py
class Solution:
    def knapsack(self, W, val, wt):
        
        n = len(val)                                 
        
        items = [[val[i], wt[i]] for i in range(n)]     
        
        items.sort(key=lambda x: x[0]/x[1], reverse=True)     
        
        res = 0.0                                   
        currentCapacity = W                        
        
        for i in range(len(items)):                 
            if currentCapacity == 0:                
                break
            
            if items[i][1] <= currentCapacity:     
                res += items[i][0]                  
                currentCapacity -= items[i][1]  
            else:
                res += (items[i][0]/items[i][1]) * currentCapacity     
                break
             
        return res                                  


class Solution:
    def knapsack(self, W, val, wt):
        n = len(val)

        # Create (value, weight) pairs
        items = [(val[i], wt[i]) for i in range(n)]

        # Sort items by value/weight ratio (descending)
        items.sort(key=lambda x: x[0] / x[1], reverse=True)

        total_value = 0.0
        current_capacity = W

        for value, weight in items:
            if current_capacity == 0:
                break

            if weight <= current_capacity:
                # Take full item
                total_value += value
                current_capacity -= weight
            else:
                # Take fractional part
                total_value += (value / weight) * current_capacity
                break

        return total_value

File: GreedyMethod/javaScript/fractionalKnapsack.js
/**
 * @param {number[]} val
 * @param {number[]} wt
 * @param {number} capacity
 * @returns {number}
 */
class Solution {
  fractionalKnapsack(val, wt, capacity) {
    let n = val.length;

    // Create (value, weight) pairs
    let items = [];
    for (let i = 0; i < n; i++) {
      items.push([val[i], wt[i]]);
    }

    // Sort by value/weight ratio (descending)
    items.sort((a, b) => b[0] / b[1] - a[0] / a[1]);

    let res = 0.0;
    let remainingCapacity = capacity;

    for (let i = 0; i < n; i++) {
      let value = items[i][0];
      let weight = items[i][1];

      if (remainingCapacity === 0) break;

      if (weight <= remainingCapacity) {
        res += value;
       res += value;
        remainingCapacity -= weight;
      } else {
        res += (value / weight) * remainingCapacity;
        break;
      }
    }

    return res;
  }
}

let val = [60, 100, 120];
let wt = [10, 20, 30];
let capacity = 50;

let sol = new Solution();
console.log(sol.fractionalKnapsack(val, wt, capacity));
